docs/pallas/pipelining.md

```diff
@@ -0,0 +1,591 @@
+  kernelspec:
+    display_name: Python 3
+    name: python3
+---
+
+<!-- #region id="C93Xlf0DRW9H" -->
+# Software Pipelining
+
+Software pipelining is an important technique in kernel writing for optimizing performance by overlapping multiple asynchronous operations even if there are data dependencies between them. In the context of kernel writing, the most common form of pipelining involves overlapping communication and memory transfers with compute such that the hardware accelerator never stalls while waiting for data to arrive. Therefore, we will solely focus on the problem of communication-compute pipelining in this tutorial. We will begin by covering the problem conceptually, outlining the Pallas API for writing pipelines, and going over some realistic examples using the API.
+
+This tutorial only covers the conceptual foundations of pipelining. For platform-specific references, please see the [TPU](https://docs.jax.dev/en/latest/pallas/tpu/pipelining.html), or GPU (coming soon!) specific pipelining references.
+
+<!-- #endregion -->
+
+```python id="YkOjspo5BKPD" executionInfo={"status": "ok", "timestamp": 1744763768352, "user_tz": 420, "elapsed": 204, "user": {"displayName": "Justin Fu", "userId": "17543197034567316452"}}
+import jax
+from jax import numpy as jnp
+from jax.experimental import pallas as pl
+import numpy as np
+```
+
+<!-- #region id="shnVghWUSvpx" -->
+## Memory Hierarchies
+
+The first step in understanding pipelining conceptually involves understanding the different forms of memory available and the tradeoffs between them. Most hardware architectures (including CPUs, GPUs, and TPUs) utilize a wide variety of memory spaces that tradeoff capcity vs latency/bandwidth. For the purpose of Pallas, we are typically interested in registers, SRAM, DRAM, and potentially network communiation.
+- **Registers** are the closest type of memory to the processor, and typically values must be loaded directly into registers before doing any compute on them.
+- **SRAM** (also known as Shared Memory/L1 and L2 cache on GPUs, or VMEM on TPUs) also lives fairly close to the processor, but has larger capacity than registers.
+SRAM on modern ML accelerators typically range in the 10-100MB range (TPU v5p contains 96MB of VMEM, and H100 GPUs contain ~30MB of L1 cache and 50MB of L2).
+It's reasonable to expect the latency to access SRAM to be on the order of 10x longer than accessing a register.
+- **DRAM** (also known as HBM) has much higher capacity than SRAM, typically in the 10-100GB range for modern ML accelerators. However, the latency is roughly on the order of 10x longer to access compared to SRAM.
+- **Network** communication becomes crucial for larger workloads when the size of DRAM on a single device becomes insufficient. We do not cover distributed pipelining in this tutorial, but see the [distributed TPU kernels](https://docs.jax.dev/en/latest/pallas/tpu/distributed.html) guide for writing pipelines across multiple devices.
```

Comment:
"when the size of DRAM on a single device becomes insufficient"

maybe add something like:
when the size of DRAM on a single device becomes insufficient or when we'd like to take advantage of parallel computations.

Reply:
Done.